### PR TITLE
`Self` in type definitions (self_in_typedefs)

### PR DESCRIPTION
**Pull request by alexreg**
This implements the [`self_in_typedefs` feature](https://github.com/rust-lang/rfcs/blob/master/text/2300-self-in-typedefs.md) ([tracking issue 49303](https://github.com/rust-lang/rust/issues/49303)).

r? @eddyb

CC @Centril

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -2161,16 +2161,18 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
...
-            ItemKind::Struct(_, ref generics) |
-            ItemKind::Union(_, ref generics) |
             ItemKind::Fn(_, _, ref generics, _) => {
                 self.with_type_parameter_rib(HasTypeParameters(generics, ItemRibKind),
-                                         |this| visit::walk_item(this, item));
+                                             |this| visit::walk_item(this, item));
             }
 
+            ItemKind::Enum(_, ref generics) |
+            ItemKind::Existential(_, ref generics) |
```
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: Yep, I'm with you here... let's bring it up again at a later point. Hopefully after your `type Foo = impl Bar` RFC is merged.

[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc_resolve/lib.rs]: (@varkor wrote that one ^,-)

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -2440,6 +2442,21 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         self.ribs[TypeNS].pop();
     }
 
+    fn resolve_type(&mut self,
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: I'd rename this to `resolve_type_definition_with_self_ty` or something more "obvious" - but why does it exists, couldn't you write this in-place?

**Comment by alexreg** [comment]
Err, is tidy broken here? That line definitely isn't longer than 100 chars.

**Comment by varkor** [comment]
@alexreg: I'm not sure about the 100 character message, but there's an issue with the diagnostic length:
```
error: description for error code E0411 contains a line longer than 80 characters.
```

**Comment by alexreg** [comment]
@varkor Yeah, I noticed that after I posted the above, but thanks!

File: src/test/ui/feature-gates/feature-gate-panic-implementation.rs [hunk 3]
```diff
@@ -8,8 +8,6 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// compile-flags:-C panic=abort
```
Comment (varkor) [inline src/test/ui/feature-gates/feature-gate-panic-implementation.rs]: Why are you modifying this test at all?

Comment (alexreg) [reply src/test/ui/feature-gates/feature-gate-panic-implementation.rs]: Accident. :-P

File: src/test/compile-fail/feature-gate-self-in-typedefs.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+enum StackList<'a, T: 'a> {
+    Nil,
+    Cons(T, &'a Self)
+    //~^ ERROR cannot find type `Self` in this scope
+    //~| `Self` is only available in traits and impls
+}
```
Comment (estebank) [inline src/test/compile-fail/feature-gate-self-in-typedefs.rs]: Could you turn this into an `ui` test?

Comment (alexreg) [reply src/test/compile-fail/feature-gate-self-in-typedefs.rs]: Yep, will do. My mistake.

Comment (estebank) [reply src/test/compile-fail/feature-gate-self-in-typedefs.rs]: No problem! It's just a nitpick.

**Comment by alexreg** [comment]
@Centril Turns out having this feature for `existential type` is useless, since it prevents it resolving (endless cycle). Makes sense, once you think about it!

**Comment by Centril** [comment]
@alexreg  leave it out ;)

File: src/librustc_resolve/diagnostics.rs [hunk 5]
```diff
@@ -770,17 +770,18 @@ match x {
...
 E0411: r##"
-The `Self` keyword was used outside an impl or a trait.
+The `Self` keyword was used outside an impl, trait, or type definition.
 
 Erroneous code example:
 
 ```compile_fail,E0411
-<Self>::foo; // error: use of `Self` outside of an impl or trait
+<Self>::foo; // error: use of `Self` outside of an impl, trait, or type
+             // definition
```
Comment (Centril) [inline src/librustc_resolve/diagnostics.rs]: Does `existential type` count as a "type definition"? (It's not *really* defining a new type...)

File: src/test/run-pass/self-in-typedefs.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(existential_type)]
```
Comment (Centril) [inline src/test/run-pass/self-in-typedefs.rs]: **If** this is going to be included there should be a test for this as well ^,-

Comment (alexreg) [reply src/test/run-pass/self-in-typedefs.rs]: I forgot to remove it. ;-)

File: src/doc/unstable-book/src/language-features/self-in-typedefs.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
+# `self_in_typedefs`
+
+The tracking issue for this feature is: [#49303]
+
+[#49303]: https://github.com/rust-lang/rust/issues/49303
+
+------------------------
+
+The `self_in_typedefs` feature gate lets you use the special `Self` identifier
+in `struct`, `enum`, `union`, and `existential type` type definitions.
```
Comment (varkor) [inline src/doc/unstable-book/src/language-features/self-in-typedefs.md]: No longer in `existential type`.

File: src/librustc_resolve/lib.rs [hunk 8]
```diff
@@ -2474,6 +2482,19 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         self.ribs[TypeNS].pop();
     }
 
+    fn resolve_type_definition_with_self_ty(&mut self,
```
Comment (varkor) [inline src/librustc_resolve/lib.rs]: Is there any reason not to inline this (with an explanatory comment)?

File: src/librustc_resolve/lib.rs [hunk 9]
```diff
@@ -2192,14 +2192,22 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
...
-            ItemKind::Struct(_, ref generics) |
-            ItemKind::Union(_, ref generics) |
             ItemKind::Fn(_, _, ref generics, _) => {
                 self.with_type_parameter_rib(HasTypeParameters(generics, ItemRibKind),
-                                         |this| visit::walk_item(this, item));
+                                             |this| visit::walk_item(this, item));
+            }
+
+            ItemKind::Enum(_, ref generics) |
+            ItemKind::Existential(_, ref generics) |
```
Comment (varkor) [inline src/librustc_resolve/lib.rs]: I thought this no longer applied to `existential type`?

Comment (alexreg) [reply src/librustc_resolve/lib.rs]: It isn't. I think you're looking at an old version of the source somehow... weird.

File: src/librustc_resolve/lib.rs (hunk 8 above)
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: It's cleaner and clearer this way, I think.

Comment (eddyb) [reply src/librustc_resolve/lib.rs]: I disagree. You could have only one `with_type_parameter_rib` (exactly where it used to be), and then only make the `with_self_rib` conditional.

File: src/librustc_resolve/lib.rs [hunk 10]
```diff
@@ -2968,7 +2990,12 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
             if is_self_type(path, ns) {
                 __diagnostic_used!(E0411);
                 err.code(DiagnosticId::Error("E0411".into()));
-                err.span_label(span, "`Self` is only available in traits and impls");
+                let available_in = if this.session.features_untracked().self_in_typedefs {
+                    "impls, traits, and type definitions"
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: The order here changed, is that intentional?

File: src/librustc_resolve/lib.rs (hunk 8 above)
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: Fair enough, I'll do that then. Just to make you happy!

File: src/librustc_resolve/lib.rs (hunk 10 above)
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: Yes, it is. I had accidentally deleted it before and rewritten it the wrong way round, which broke some UI tests. :-P So this fixes it.

**Review by eddyb — APPROVED** [review]

**Pull request merged**